qiskit/circuit/library/grover_operator.py

```diff
@@ -0,0 +1,196 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The Grover operator."""
+
+from typing import List, Optional, Union
+import numpy
+from qiskit.circuit import QuantumCircuit, QuantumRegister, AncillaRegister
+from qiskit.quantum_info import Statevector, Operator, DensityMatrix
+from .standard_gates import MCXGate
+
+
+class GroverOperator(QuantumCircuit):
+    r"""The Grover operator.
+
+    Grover's search algorithm [1, 2] consists of repeated applications of the so-called
+    Grover operator used to amplify the amplitudes of the desired output states.
+    This operator consists of the oracle, $\mathcal{S}_f$, that multiplies the amplitude of the
+    good states by -1, a reflection $\mathcal{S}_0$ about the $\ket{0}^{\otimes n}$ state
+    and an input state $\mathcal{A}$. For the textbook Grover search, $\mathcal{A} = H^{\otimes n}$,
+    however for generic amplitude amplification the input state might differ [3].
+    With these terms, the Grover operator can be written as
+
+    .. math::
+
+        \mathcal{Q} = \mathcal{A} \mathcal{S}_0 \mathcal{A} \mathcal{S_f}
+
+    .. note::
+
+        Sometimes the Grover operator is defined with a negative sign, in that case
+        $\mathcal{S}_0$ multiplies all states *except* $\ket{0}^{\otimes n}$ with -1.
+        In our formulation, $\mathcal{S}_0$ only multiplies $\ket{0}^{\otimes n}$ with -1.
+
+    Examples:
```

Comment:
I would add an example that uses all parameters.

Reply:
Added in b371c73 (shows all params except zero_reflection and reflection_qubits, since they are mutually exclusive). Also added one with density matrices/statevector.